src/server/serveractiveobject.cpp

```diff
@@ -67,31 +73,78 @@ std::string ServerActiveObject::generateUpdateInfantCommand(u16 infant_id, u16 p
 	return os.str();
 }
 
-void ServerActiveObject::dumpAOMessagesToQueue(std::queue<ActiveObjectMessage> &queue)
+void SAO::dumpAOMessagesToQueue(std::queue<ActiveObjectMessage> &queue)
 {
 	while (!m_messages_out.empty()) {
 		queue.push(std::move(m_messages_out.front()));
 		m_messages_out.pop();
 	}
 }
 
-void ServerActiveObject::markForRemoval()
+void SAO::markForRemoval()
 {
 	if (!m_pending_removal) {
 		onMarkedForRemoval();
 		m_pending_removal = true;
 	}
 }
 
-void ServerActiveObject::markForDeactivation()
+void SAO::markForDeactivation()
 {
 	if (!m_pending_deactivation) {
 		onMarkedForDeactivation();
 		m_pending_deactivation = true;
 	}
 }
 
-InventoryLocation ServerActiveObject::getInventoryLocation() const
+InventoryLocation SAO::getInventoryLocation() const
 {
 	return InventoryLocation();
 }
+
+void SAO::invalidateFinalObservers()
+{
+	m_final_observers.reset();
+}
+
+const SAO::Observers &SAO::getFinalObservers()
+{
+	if (m_final_observers) {
+		return *m_final_observers;
+	}
+	auto parent = getParent();
+	if (parent == nullptr) {
+		return *(m_final_observers = m_observers);
+	}
+	auto parent_observers = parent->getFinalObservers();
+	if (!parent_observers) {
+		return *(m_final_observers = m_observers);
+	}
+	if (!m_observers) {
+		return *(m_final_observers = parent_observers);
+	}
+	// Set intersection between parent_observers and m_observers
+	m_final_observers = std::unordered_set<std::string>(
+			std::min(m_observers->size(), parent_observers->size()));
```

Comment:
the idea was that this could be a reasonable hash table size hint to avoid unnecessary reallocations. given that this is only an upper bound however, and the hash table might want to be "denser" than this, this was probably a bad idea.